### DISCUSSION TITLE
Resolving conflict help!

### DISCUSSION THREAD

**Discussion by Vextamour** [Mimetis/Dotmim.Sync#733] [Q&A]
I have a WPF application that is always reading from the SQLite databases online or offline. Then anytime they navigate in the app or save anything I'm always making sync back calls through a WebAPI to keep the SQLite data up to date to the cloud Azure SQL Server. The issue I'm having is resolving conflicts in multiple scenarios below. Ultimately, I think I'm seeking help on how to resolve conflicts and chose what data to update or insert and sync back to the server. I've read documentation but I'm just not getting it to function properly. _At this point I'm not sure if it's allowed but I'd even be happy to pay for some 1:1 time because I need to get past this hold-up quickly. I've been scavenging through Issues and Discussions on here trying to find answers to no avail. I'm on DotMimSync version 0.9.4. Thank you kindly to anyone who can help me 🙏 !

One scenario is a basic ComputerA input data and ComputerB syncs up to the cloud. However, if I edit a row of data on ComputerB it raises a conflict on the client but never updates the cloud server. I've tried the two sync trick with no consistent luck (also I don't want to specify every single column by name) and attached my TWO SYNC code to say take the client edit over the server (code from local application below). If I step through the code I can see in LocalRow a value change. Then I see it update the RemoteRow value change but then it just never makes it to the server. Later down the road I would like to prompt the user asking which version of data they want but for now I need to get this just syncing at least. 

Second scenario there are two computers running the application. ComputerA inserts or updates data then syncs up to the cloud while the person on ComputerB is still filling out an interface. Then ComputerB adds a row but it raises a conflict because the server says the row doesn't exist so it inserts into the SQLite database syncs up to the server and since it doesn't exist there it gets removed instantly out of local database. 

```
                string clientDBConnectionString = Globals.SQLLiteConnectionString;
                string baseUrl = ConfigurationManager.AppSettings["BaseURL"];

                HttpClient httpClient = new HttpClient();
                httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", Globals.Auth0Token.Replace("Bearer ", ""));

                // Adding the HttpClient instance to the web client orchestrator
                var serverOrchestrator = new WebClientOrchestrator($"{baseUrl}api/sync", client: httpClient);

                //var serverOrchestrator = new WebClientOrchestrator("https://localhost:44368/api/sync"); //Uncomment this if you're trying to test the API sync locally
                var clientProvider = new SqliteSyncProvider(clientDBConnectionString);

                var options = new SyncOptions { ConflictResolutionPolicy = ConflictResolutionPolicy.ServerWins };
                
                var agent = new SyncAgent(clientProvider, serverOrchestrator, options);

                var localOrchestrator = agent.LocalOrchestrator;
                var remoteOrchestrator = agent.RemoteOrchestrator;

                try
                {

                    agent.LocalOrchestrator.OnApplyChangesFailed(acf =>
                    {
                        var localRow = acf.Conflict.LocalRow; //Local machine SQLite
                        var remoteRow = acf.Conflict.RemoteRow; //Azure SQL Server

                        //I don't want to have to call each row by name like this. I'd rather call it dynamically in a loop of sorts.
                        //Also in the future I'd like the user to pick but for now just trying to get the client to override what's going to the cloud server.
                        remoteRow["WellNumber"] = localRow["WellNumber"];

                        //I see the remote row updated the WellNumber based off the SQLite local row. 

                        //acf.FinalRow = localRow; /** TRIED FORCING THE FINAL ROW TO LOCAL ROW DIDN'T WORK EITHER**/
                        acf.SenderScopeId = null;
                    });

                    var results = await agent.SynchronizeAsync();

                        MessageBox.Show($"Sync duration: {results.CompleteTime.Subtract(results.StartTime).TotalSeconds}\n" +
                        $"Changes: {results.TotalChangesUploaded}\n" +
                        $"Downloaded: {results.TotalChangesDownloaded}\n" +
                        $"Applied: {results.TotalChangesApplied}\n" +
                        $"Conflicts: {results.TotalResolvedConflicts}");

                    results = await agent.SynchronizeAsync();

                    MessageBox.Show($"Sync duration: {results.CompleteTime.Subtract(results.StartTime).TotalSeconds}\n" +
                                    $"Changes: {results.TotalChangesUploaded}\n" +
                                    $"Downloaded: {results.TotalChangesDownloaded}\n" +
                                    $"Applied: {results.TotalChangesApplied}\n" +
                                    $"Conflicts: {results.TotalResolvedConflicts}");
```

**Comment by Vextamour** [comment]
@Mimetis  I see you're active a lot on discussions and issues. Anyway possible you could take a look at this and provide feedback please? Or anyone else. My issue ticket has been open for 14 days and this one for 6 days and really need to progress on a project. 

Just curious too I came across Microsoft.Datasync.Client NuGet package are you responsible for this too because it seems very familiar to what you're already doing.

**Reply by Mimetis** [reply]
I did not respond, because I don't understand your problem actually.
Please try to provide a sample repro, to allow me to understand your problem

**Reply by Vextamour** [reply]
@Mimetis I honestly don't even know how to provide a sample repo. I have DotMim creating local SQLite databases and creating all the tables/populating with data on any client that app runs on. I'm able to insert records on different machines. However, my first issue is if one computer inserts the data only that computer can edit the data and have it get pushed to the Azure SQL Server. No other client is able to update that row and push the changes to the cloud server. If I try it creates a conflict which I'm struggling to resolve and push changes. I've tried you two sync method you mentioned to no avail. If you look at code above you can see near the bottom where I'm making the two sync call back. Could you provide a sample perhaps on how you resolve conflicts and maybe I can go off that to fix up my issues please?

**Reply by Mimetis** [reply]
In your table, what kind of primary key are you using ? Identity auto increment ? GUID ?

**Reply by Vextamour** [reply]
Truly appreciate your help! All my tables have a primary key using int identity(1,1) on Azure SQL Server. I'm not doing anything additional when DotMim creates the SQLite database except creating views. Below is a sample of a table I'm creating. 

```
create table [dbo].[Wells]
(
    WellID int identity(1,1) not null,
	CompanyID int not null,	
	WellTypeID int not null,
	WellConfigurationTypeID int not null,
	StateID int,
	StartDate datetime,
	EndDate datetime,
	API varchar(500),
	PropertyNumber varchar(50),
	WellName varchar(5000),
	WellNumber varchar(100),
	County varchar(5000),
	Longitude decimal (12,9),
	Latitude decimal (12,9),
	UnitLetter varchar(100),
	Section varchar(100),
	Township varchar(1000),
	[Range] varchar(100),
	PumpDescription varchar(5000),
	FieldName varchar(500),
	Elevation int,
	KBTubingHeadDistance decimal(8,1),
	SpudDate datetime,
	FinalDrigDate datetime,
	TotalMeasuredDepth int,
	TotalVerticalDepth int,
	PlugBackTotalDepth int

    constraint PKC_Wells_WellID primary key clustered (WellID) on [PRIMARY]
);

--Below are FKC and Indexes didn't paste here
```

**Reply by Mimetis** [reply]
That's why you have an issue
DONT use identity auto increment primary keys

If you are creating two record on two separate databases, with the same ID, how can you merge these 2 records on the server ?

The fact is that the first database that make the sync will correctly upload the record to the server with the id
The second database will upload its record afterwards, but server will see it as an update, and will generate a conflict (and will send back the first version, since server has the truth in case of conflict)

Take a look here https://github.com/Mimetis/Dotmim.Sync/issues/590

And use GUID for your primary keys

**Reply by Vextamour** [reply]
I just read that documentation and I'm confused on how all this ties together. 

When you say use GUID for my primary keys are you simply suggesting what I changed down below by going from int identity(1,1) to **WellID UNIQUEIDENTIFIER PRIMARY KEY default NEWID() not null** as my new primary key?

Otherwise, are you suggesting from documentation to create a scope_info_identity table and anytime a new client downloads the application I have to insert that client information and add tables with min/max identity ranges and manage those? 

```
create table [dbo].[Wells]
(
        WellID UNIQUEIDENTIFIER PRIMARY KEY default NEWID() not null,
	CompanyID int not null,	
	WellTypeID int not null,
	WellConfigurationTypeID int not null,
	StateID int,
	StartDate datetime,
	EndDate datetime,
	API varchar(500)
       --Additional columns here

    constraint PKC_Wells_WellID primary key clustered (WellID) on [PRIMARY]
);

--Below are FKC and Indexes didn't paste here
```

**Reply by Mimetis** [reply]
- Use GUID
- Restart everything from scratch (or remove all stored proc / triggers / tracking tables by calling `DeprovisionAsync` )
- Test

**Reply by Vextamour** [reply]
Thanks Mimetis. I have changed all my tables to use uniqueidentifers and in my C# application I'm currently going through changing everything to using Guid. I already deprovisioned and then provisioned the Azure SQL Server. Once I finish fixing all the errors in my C# application I will test the sync process again and report back. Thank you again for your response and help! Really hoping this fixes everything up.

**Reply by Vextamour** [reply]
@Mimetis 
1. I've changed entire Azure DB using uniqueidentifers and updated entire C# application using Guid instead of int for primary keys.
2. Deprovisoned my Azure SQL Server and then provisioned SQL Server.
3. **Testing failed** to update. I inserted a record from Laptop. Ran application on Desktop to update the new row and it updates the local SQLite database recognizes it as a conflict and overwrites SQLite database with server version instantly. I tried making 2 call sync as suggested from documentation (code below) but it doesn't appear to do anything. Any other suggestions would be much appreciated!!!

First sync shows this so I can see it coming through as a conflict. 
Sync attempt 1 output:
Changes: 1
Downloaded: 2
Applied: 2
Conflicts: 1

2 sync process output shows it resolved the conflict since I have ServerWins so it just wiped out my new value that I want to push to the server.
Sync attempt 2 output:
Changes: 0
Downloaded: 2
Applied: 2
Conflicts: 0

```
            //This is entire sync call from application
                if(Globals.Auth0Token == null || Globals.Auth0Token == "")
                    Globals.Auth0Token = GenericMethodHelpers.GetAuth0Token();

                string clientDBConnectionString = Globals.SQLLiteConnectionString;
                string baseUrl = ConfigurationManager.AppSettings["BaseURL"];

                HttpClient httpClient = new HttpClient();
                httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", Globals.Auth0Token.Replace("Bearer ", ""));

                // Adding the HttpClient instance to the web client orchestrator
                var serverOrchestrator = new WebClientOrchestrator($"{baseUrl}api/sync", client: httpClient);

                //var serverOrchestrator = new WebClientOrchestrator("https://localhost:44368/api/sync"); //Uncomment this if you're trying to test the API sync locally
                var clientProvider = new SqliteSyncProvider(clientDBConnectionString);

                var options = new SyncOptions { ConflictResolutionPolicy = ConflictResolutionPolicy.ServerWins };
                
                var agent = new SyncAgent(clientProvider, serverOrchestrator, options);

                var localOrchestrator = agent.LocalOrchestrator;
                var remoteOrchestrator = agent.RemoteOrchestrator;

                try
                {
                   //DOUBLE SYNC BACK CALL HERE
                    var results = await agent.SynchronizeAsync();

                    MessageBox.Show($"Sync duration: {results.CompleteTime.Subtract(results.StartTime).TotalSeconds}\n" +
                    $"Changes: {results.TotalChangesUploaded}\n" +
                    $"Downloaded: {results.TotalChangesDownloaded}\n" +
                    $"Applied: {results.TotalChangesApplied}\n" +
                    $"Conflicts: {results.TotalResolvedConflicts}");

                    results = await agent.SynchronizeAsync();

                    MessageBox.Show($"Sync duration: {results.CompleteTime.Subtract(results.StartTime).TotalSeconds}\n" +
                                    $"Changes: {results.TotalChangesUploaded}\n" +
                                    $"Downloaded: {results.TotalChangesDownloaded}\n" +
                                    $"Applied: {results.TotalChangesApplied}\n" +
                                    $"Conflicts: {results.TotalResolvedConflicts}");
                    MessageBox.Show("DB Sync Successful");
                    return true;
             
```

```
//SAMPLE OF A TABLE
create table [dbo].[CompanyEmployees]
(
       CompanyEmployeeID uniqueidentifier default newid() not null,
	CompanyID uniqueidentifier not null,
	StateID uniqueidentifier,
	IndividualRelationTypeID uniqueidentifier,
	EmployeeLocalNumber varchar(100),
	FirstName varchar(1000),
	MiddleName varchar(1000),
	LastName varchar(1000),

    constraint PKC_CompanyEmployees_CompanyEmployeeID primary key clustered (CompanyEmployeeID) on [PRIMARY]
);
```

**Reply by Mimetis** [reply]
Here is my setup:
2 projects:
- 1 console application (just faster for me, but would be the same if it's a Xamarin app or WPF or any kind of client application)
- 1 server ASP.NET application (.NET 6.0) to handle the sync process

I'm making a SQL initialization, using this script to create the server database (**AdventureWorks**) and client (empty) database (**Client**)

``` sql
USE [master]
GO 

-- Server database
if (exists (select * from sys.databases where name = 'AdventureWorks'))
Begin
	ALTER DATABASE [AdventureWorks] SET  SINGLE_USER WITH ROLLBACK IMMEDIATE;
	DROP DATABASE [AdventureWorks]
End
Create database [AdventureWorks];

-- Client database
if (exists (select * from sys.databases where name = 'Client'))
Begin
	ALTER DATABASE [Client] SET  SINGLE_USER WITH ROLLBACK IMMEDIATE;
	DROP DATABASE [Client]
End
Create database [Client];


use AdventureWorks;

create table [dbo].[CompanyEmployees]
(
       CompanyEmployeeID uniqueidentifier default newid() not null,
	CompanyID uniqueidentifier not null,
	StateID uniqueidentifier,
	IndividualRelationTypeID uniqueidentifier,
	EmployeeLocalNumber varchar(100),
	FirstName varchar(1000),
	MiddleName varchar(1000),
	LastName varchar(1000),

    constraint PKC_CompanyEmployees_CompanyEmployeeID primary key clustered (CompanyEmployeeID) on [PRIMARY]
);


-- Inserting a first row
Insert into CompanyEmployees Values (newid(), newid(), newid(), newid(), 'A001', 'John', '', 'Doe')
```

The server is using the classic web sync server configurtion:

`Program.cs`
``` cs
var builder = WebApplication.CreateBuilder(args);
builder.Services.AddDistributedMemoryCache();
builder.Services.AddSession(options => options.IdleTimeout = TimeSpan.FromMinutes(30));

// [Required]: Get a connection string to your server data source
var connectionString = builder.Configuration.GetSection("ConnectionStrings")["SqlConnection"];
// [Required] Tables involved in the sync process:
var tables = new string[] { "CompanyEmployees" };
// [Required]: Add a SqlSyncProvider acting as the server hub.
builder.Services.AddSyncServer<SqlSyncProvider>(connectionString, tables);

builder.Services.AddControllers();
builder.Services.AddEndpointsApiExplorer();
builder.Services.AddSwaggerGen();

var app = builder.Build();

app.UseSwagger();
app.UseSwaggerUI();
app.UseHttpsRedirection();
app.UseAuthorization();
app.UseSession();
app.MapControllers();
app.Run();
```

`SyncController.cs`:
``` cs
[ApiController]
[Route("api/[controller]")]
public class SyncController : Controller
{
    private WebServerOrchestrator orchestrator;
    private readonly IWebHostEnvironment env;

    public SyncController(WebServerOrchestrator webServerOrchestrator, IWebHostEnvironment env)
    {
        this.orchestrator = webServerOrchestrator;
        this.env = env;
    }
    [HttpPost]
    public Task Post()
        => orchestrator.HandleRequestAsync(this.HttpContext);

    [HttpGet]
    public async Task Get()
    {
        if (env.IsDevelopment())
            await WebServerOrchestrator.WriteHelloAsync(this.HttpContext, orchestrator);
    }
}
```

And now my test from the Console application (see comments for results):

``` cs
// Sync 2 clients
var serverProvider = new WebClientOrchestrator("https://localhost:7293/api/sync");
var clientProvider = new SqlSyncProvider(clientConnectionString);

var agent = new SyncAgent(clientProvider, serverProvider);

// 1) First Step: Initial sync
var result = await agent.SynchronizeAsync();

Console.WriteLine(result);
// Synchronization done.
// Total changes  uploaded: 0
// Total changes  downloaded: 1
// Total changes  applied: 1
// Total resolved conflicts: 0


// 2) Step. Insert into local database
Guid companyEmployeeID = Guid.NewGuid();
using (var c = new SqlConnection(clientProvider.ConnectionString))
{
    c.Open();
    var commandText = @"Insert into CompanyEmployees 
            (CompanyEmployeeID, CompanyID, StateID, IndividualRelationTypeID, 
                EmployeeLocalNumber, FirstName, MiddleName, LastName)
            Values (@CompanyEmployeeID, @CompanyID, @StateID, @IndividualRelationTypeID,
                @EmployeeLocalNumber, @FirstName, @MiddleName, @LastName)";

    var command = new SqlCommand(commandText, c);
    command.Parameters.AddWithValue("@CompanyEmployeeID", companyEmployeeID);
    command.Parameters.AddWithValue("@CompanyID", Guid.NewGuid());
    command.Parameters.AddWithValue("@StateID", Guid.NewGuid());
    command.Parameters.AddWithValue("@IndividualRelationTypeID", Guid.NewGuid());
    command.Parameters.AddWithValue("@EmployeeLocalNumber", "A002");
    command.Parameters.AddWithValue("@FirstName", "Ewan");
    command.Parameters.AddWithValue("@MiddleName", "Mac");
    command.Parameters.AddWithValue("@LastName", "Gregor");
    command.ExecuteNonQuery();

    c.Close();
}

// 3) Make a sync to upload the new record to server
result = await agent.SynchronizeAsync();

Console.WriteLine(result);
// Synchronization done.
// Total changes  uploaded: 1
// Total changes  downloaded: 0
// Total changes  applied: 0
// Total resolved conflicts: 0


// 4) Make a local update on this particular row
using (var c = new SqlConnection(clientProvider.ConnectionString))
{
    c.Open();
    var commandText = "Update CompanyEmployees set LastName = @LastName Where CompanyEmployeeID = @CompanyEmployeeID";

    var command = new SqlCommand(commandText, c);
    command.Parameters.AddWithValue("@CompanyEmployeeID", companyEmployeeID);
    command.Parameters.AddWithValue("@LastName", "Gregor Obiwan Kenobi");
    command.ExecuteNonQuery();

    c.Close();
}

// 5) Make a sync to upload the updated record to server
result = await agent.SynchronizeAsync();

Console.WriteLine(result);
// Synchronization done.
// Total changes  uploaded: 1
// Total changes  downloaded: 0
// Total changes  applied: 0
// Total resolved conflicts: 0

// 6) Make a server update on this particular row
using (var c = new SqlConnection(serverConnectionString))
{
    c.Open();
    var commandText = @"Update CompanyEmployees set FirstName = @FirstName, 
                        MiddleName = @MiddleName, 
                        LastName = @LastName 
                        Where CompanyEmployeeID = @CompanyEmployeeID";

    var command = new SqlCommand(commandText, c);
    command.Parameters.AddWithValue("@CompanyEmployeeID", companyEmployeeID);
    command.Parameters.AddWithValue("@FirstName", "Obiwan");
    command.Parameters.AddWithValue("@MiddleName", "");
    command.Parameters.AddWithValue("@LastName", "Kenobi");
    command.ExecuteNonQuery();

    c.Close();
}

// 7) Make a sync to download the updated record from the server
result = await agent.SynchronizeAsync();

Console.WriteLine(result);
// Synchronization done.
// Total changes  uploaded: 0
// Total changes  downloaded: 1
// Total changes  applied: 1
// Total resolved conflicts: 0

// The 2 lines ar strictly identitcal
```
<img width="569" alt="image" src="https://user-images.githubusercontent.com/4592555/170765741-24240fcd-fd34-457e-9dca-20f0f9a0dde4.png">

Now you have a walkthrough, please use it to reproduce your issue, will be easier to help you

**Reply by Vextamour** [reply]
Sorry for delayed response. I've tried all your steps and here is my narrowed down issue but I'm still not sure how you solve. 

1. I deprovisioned/provisioned SQL Server again. Deleted local SQLite database. I run application with SQL Server connection and it downloads/builds SQLite database with all server data on client. 

2. I insert a row into server and call sync and it pushes row to client SQLite. I update that row it updates on server calls sync and pushed to client as well. All is good so far. 

3. I change connection string to read from SQLite now. If I insert a _row it adds to client SQLite_ calls sync and pushes to server perfect. **HOWEVER, my issue still is if I update the original row inserted from step 2 it updates the client SQLite database calls sync and then overrides it with the server value because it see's the conflict. I do have ServerWins enabled by default still but I want the server to be updated by the client values in this case. I tried the two sync process back to back from documentation and that didn't take the local client values.** What do I need to do differently to get sync process to take the client changes to push to server?

This is output after it tries to push the SQLite changed row to try to push to Server.
```
// Synchronization done.
// Total changes  uploaded: 1
// Total changes  downloaded: 2
// Total changes  applied: 2
// Total resolved conflicts: 1
```

**Reply by Mimetis** [reply]
Take my sample, modify the code to reproduce your issue, and share it

**Reply by Vextamour** [reply]
Thanks I'll replicate and report back. Started a new job so my side business development is getting backlogged sadly. Appreciate your help!

**Reply by Vextamour** [reply]
I got everything replicated, however, this is going from SQL Server DB to SQL Server DB. It worked testing between 3 separate databases I made on local as (separate clients). What tweaks do I need to make to have it go SQL Server (Main Server) to SQLite because that is where my issue is occurring.

**Reply by Mimetis** [reply]
hum, use a `SqliteSyncProvider` instead of a `SqlSyncProvider` ? (and adapt the queries using `SqliteConnection` to reproduce the user queries)

**Reply by Vextamour** [reply]
I'm still working on this. I started a new job and have made many modifications to my code base for this project. I'm almost to a point where I can focus on getting this functionality to test again. Thank you for the help!
